File: Leetcode/121. Best Time to Buy and Sell Stock.ts
function maxProfit(prices: number[]): number {
    let arr: number[] = [];
    let n: number = prices.length;
    arr[n-1] = prices[n-1];
    for(let i=n-2; i>=0; i--) {
        arr[i] = Math.max(prices[i],arr[i+1]);
    }
    
    let ans=0;
    for(let i=0; i<n; i++) ans = Math.max(ans, arr[i]-prices[i]);
    
    return ans;
};

File: Leetcode/121. Best Time to Buy and Sell Stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int mnPrice = INT_MAX, mxProfit = INT_MIN;
        
        for(int i=0; i<prices.size(); i++) {
            mnPrice = min(mnPrice, prices[i]);
            mxProfit = max(mxProfit, prices[i] - mnPrice);
        }
        
        return mxProfit;
    }
};
